Enhancements for Thread Dump analyzer

jutzig opened the pull request:
* blocked threads visualization
 * comparing a series of thread dumps
 * parser support for thread dumps created via MXBean
 * parser support for thread dumps that start with a PID
 * visualizing CPU utilization
 * search function for 1-n thread dumps
 * diagnoser to identify typical problems
 * support for list params
 * Params can now have a default value

y1yang0 commented on backend/thread-dump-analyzer/src/main/java/org/eclipse/jifa/tda/diagnoser/ThreadDumpDiagnoser.java:
```diff
@@ -0,0 +1,138 @@
+        snapshot.getDeadLockThreads().forEach(deadlockThreads::addAll);
+        if(deadlockThreads.size() > 0) {
+            String message = String.format("%s threads are in a deadlock", deadlockThreads.size());
+            String suggestion = "Deadlocks if two or more threads are waiting for each other indefinetly. They are caused by incorrect ordering of resource locking.";
```
Maybe we can remove message and suggestion fields? I think we'd better come up with suggestions in frontend, which facilitates localization.

y1yang0 commented on frontend/src/i18n/en.js:
```diff
@@ -431,6 +433,8 @@ exports.default = {
 
     threadDump: {
       title: "Thread Dump Analysis",
+      navigation: "Navigation",
```
Counterpart in cn.js
```js
navigation: '导航',
navToTop: 'Top',
```

y1yang0 commented on frontend/src/i18n/en.js:
```diff
@@ -448,11 +452,69 @@ exports.default = {
       gcThread: "GC Thread",
       otherThread: "Other Thread",
       total: "Total",
+      monitors: "Java Monitors",
```
Counterpart in cn.js
```js
 monitors: "Java同步器",
```

y1yang0 commented on frontend/src/i18n/en.js:
```diff
@@ -448,11 +452,69 @@ exports.default = {
+      searchOptionThreadStatesPlaceholder: "Any",
+    },
+  },
+}
```
Counterpart in en.js
```
  blockedThreadsLabel: "阻塞现场",
      blockedThreads: {
        title: "{blocker} 阻塞了另一个线程 | {blocker} 阻塞了 {count} 线程"
      },
      cpuConsumingThreadsLabel: "CPU耗时",
      cpuConsumingThreads: {
        title: "展示最高耗时的线程",
        cpuConsumptionLabel: "CPU耗时",
        hours: "小时",
        minutes: "分钟",
        seconds: "秒",
        milliseconds: "毫秒",
      },
      threadDumpCompare: '线程栈比较',
      threadDumpFilePlaceholder: '填充另一个线程转储的analys页面的URL',
      diagnosis: {
        title: "诊断",
        examine: "检查",
        messageColumn: "消息",
        fileColumn: "文件",
        suggestionColumn: "建议",
      },
    },

    threadDumpCompare: {
      title: "线程栈比较",
      navigation: "导航",
      navToTop: "Top",
      stateCompare: "线程状态比较",
      stateCompareDescription: "展示多个线程栈中线程状态的分布",
      threadGroupCompare: "线程池比较",
      threadGroupCompareDescription: "如果不同线程栈文件中每个线程池中的线程数量变化较大，该线程池需要留意",
      cpuConsumingTitle: "CPU耗时线程",
      cpuConsumingDescription: "展示多个线程栈中最高CPU耗时线程。线程通过tid匹配，线程名可能在多个线程栈中改变",
      cpuConsumingDatasetLabel: "CPU耗时{unit}",
      diagnosisTitle: "诊断",
      diagnosisDescription: "提供已发现问题概览",
    },

    threadDumpSearch: {
      loadingPlaceholder: "搜索...",
      menuThreadsTitle: "线程", 
      searchTitle: "搜索",  
      resultsHeader: "1 线程匹配 | {count} 匹配",
      threadStatesChartTitle: "线程状态查询结果",
      advancedToggle: "显示高级",
      searchFields: "搜索",
      searchOptions: "搜索选项",
      searchInput: "搜索关键字",     
      searchFieldName: "线程名",
      searchFieldState: "线程状态",
      searchFieldStack: "线程调用栈",
      searchOptionRegex: "正则表达式",
      searchOptionMatchCase: "匹配大小写",
      searchOptionThreadStates: "线程状态",
      searchOptionThreadStatesPlaceholder: "任意",
    },
  },
}
```

jutzig commented:
Thanks a lot for the review and the missing translations.

Regarding the messages/suggestions for the diagnosis, I get your point, so I moved all messages to the frontend for now.
 
Internally we feed Jifa primarily via REST API and the downside of this approach (same with the GC analyzer), is that a caller of the REST API has to do quite a few steps to make the messages readable again.
Independent of this pull request, maybe it would make sense to do I18N in the backend as well, and allow that the caller provides a target locale in the REST service instead?

D-D-H commented on backend/thread-dump-analyzer/src/test/resources/jstack_11_large_with_blocked.log:
This log file contains too many lines. How about generating it before the related test cases are running?

D-D-H commented:
> Internally we feed Jifa primarily via REST API and the downside of this approach (same with the GC analyzer), is that a caller of the REST API has to do quite a few steps to make the messages readable again. Independent of this pull request, maybe it would make sense to do I18N in the backend as well, and allow that the caller provides a target locale in the REST service instead?

Agree. If the server returns any text(e.g. error message) that should be viewed by users directly, the text should be generated according to the user (caller)'s locale rather than hard code.

y1yang0 commented on frontend/src/i18n/en.js:
counterpart in cn.js
```
          DEADLOCK: "{count} 线程死锁",
          DEADLOCK_SUGGESTION: "发生死锁, 如果两个以上线程不确定性的等待对方，那么可能因为加锁顺序错误发生发生死锁",
          HIGH_BLOCKED_THREAD_COUNT: "一个线程阻塞 | {count} 个线程阻塞",
          HIGH_BLOCKED_THREAD_COUNT_SUGGESTION: "大量线程阻塞意味着系统存在瓶颈。请检查调用栈和同步加锁等代码,
          HIGH_THREAD_COUNT: "高负载线程数量{count} ",
    HIGH_THREAD_COUNT_SUGGESTION: "如此高的线程数可能导致内存耗尽和线程饥饿。查找线程泄漏或考虑使用 ThreadPools 减少线程创建并限制其数量。",
          HIGH_STACK_SIZE: "{name}有一个非常大的堆栈（> {threshold}）|{count}线程有一个非常大的堆栈大小（> {threshold}）",
          HIGH_STACK_SIZE_SUGGESTION: "堆栈过大可能导致堆栈溢出错误并降低性能。检查递归是否过深",
          HIGH_CPU_RATIO:"{name}的 CPU 占用率较高|{count}线程的 CPU 占用率较高",
          HIGH_CPU_RATIO_SUGGESTION: "如果某个线程的 CPU 比率很高，说明该线程在整个生命周期中消耗了大量 CPU。这不一定是坏事，但标志着线程非常活跃",
          THREAD_THROWING_EXCEPTION:"{name}抛出异常 | {count}线程抛出异常",
          THREAD_THROWING_EXCEPTION_SUGGESTION:"抛出异常的线程可能表明应用程序中存在问题。请记住，创建堆栈跟踪的成本很高，因此如果应用程序经常抛出异常，就会影响性能。",
```

D-D-H commented on frontend/src/components/threaddump/BlockedThreads.vue:
```diff
@@ -0,0 +1,220 @@
+
+};
+</script>
+<style>
```
Why not use Scoped CSS?

https://vue-loader.vuejs.org/guide/scoped-css.html#scoped-css

jutzig replied:
sorry, first time that I'm using Vue... Thanks for the hint, I switched it to scoped now

D-D-H commented:
This is a big enhancement, a detailed review is obviously time-consuming.

I finished a rough review and didn't find serious problems. So we consider merging it next week if there is no against.

jutzig commented:
> This is a big enhancement, a detailed review is obviously time-consuming.

Of course, I get it. Reviewing a large change like this is a lot of effort. Take all the time you need, it's no rush.

stefanroeck reviewed (commented):
In general, the usage of comparators is not consistent in this change. Personally, I prefer `===` over `==` for more strict checks.
Also, the usage of semicolons is not consistent, sometimes they are used, very often not. As there are subtle situations where missing semicolons can lead to unexpected interpretations, I'd suggest to always use them. But might be a matter of personal taste, too...

D-D-H commented:
> Personally, I prefer === over == for more strict checks.

+ 1

>  As there are subtle situations where missing semicolons can lead to unexpected interpretations, I'd suggest to always use them.

Agree. Semicolon is also friendly to compression.
Although Jifa's current front-end code doesn't have semicolons in most places.  Maybe we can enforce it in the new front-end implementation.

D-D-H commented on frontend/src/components/menu/AnalysisResultMenu.vue:
```diff
@@ -52,9 +52,23 @@
+        prefix-icon="el-icon-search"
+        size="small"
+      />
+    </b-nav-item>
```
This item causes other items to align incorrectly.

The following change should work:
```
    <b-nav-form v-if="analysisState === 'SUCCESS' && searchEnabled.includes(type)" @submit.stop.prevent=''>
      <el-input
        clearable
        v-model="searchInput"
        :placeholder="$t('jifa.threadDumpSearch.searchTitle')"
        prefix-icon="el-icon-search"
        size="small"
        @change="$emit('search', searchInput)"
      />
    </b-nav-form>
```

D-D-H commented on frontend/src/components/threaddump/ThreadDumpSearchForm.vue:
```diff
@@ -0,0 +1,152 @@
+                            <el-col :span="6">
+                                <el-button type="success" icon="el-icon-search"
+                                    @click="submitSearchForm()">{{ $t('jifa.threadDumpSearch.searchTitle') }}</el-button>
+                                <el-checkbox-button style="margin-left: 0px; " border icon="el-icon-edit"
```
el-checkbox-button should be put inside el-checkbox-group tag.

How about el-checkbox?

D-D-H reviewed (commented):
![image](https://github.com/eclipse/jifa/assets/33491035/ae25b137-65be-4073-b854-856f951e3955)

There are some errors, please fix them.

D-D-H commented on backend/thread-dump-analyzer/src/main/java/org/eclipse/jifa/tda/diagnoser/ThreadDumpAnalysisConfig.java:
```diff
@@ -0,0 +1,44 @@
+    private double highCpuConsumedRatio = 0.5;
+
+    /** issue a warning if a thread is currently throwing an exception */
+    boolean reportThrowingException = true;
```
private?

D-D-H commented on frontend/src/components/threaddump/Diagnose.vue:
```diff
@@ -0,0 +1,137 @@
+</script>
+<style>
+  
+  .el-table .cell {
```
It will influence all el-tables used by other components if not scoped style

jutzig commented:
Thanks for the review and the pointers, I added a new commit.
The error was caused by the previous commit that moved the messages  from the backend to the frontend. I forgot to handle the case of 'no issues', so thanks for spotting that. Please check the Chinese translation I added for this, since it's auto translated.

D-D-H commented on frontend/src/components/threaddump/ThreadDumpCompare.vue:
```diff
@@ -0,0 +1,417 @@
+    }
+  }
+  </script>
+  <style>
```
missing scoped here

D-D-H commented on frontend/src/components/threaddump/ThreadDumpSearch.vue:
```diff
@@ -0,0 +1,279 @@
+    }
+  }
+  </script>
+  <style>
```
missing scoped here

D-D-H reviewed (approved):
Other parts look good to me.

D-D-H commented:
If no one objects, I will merge this patch tomorrow.

Thank you for the great work! @jutzig

jutzig commented:
Thanks a lot for the helpful reviews and the merge! :+1:
